conbench/entities/commit.py

```diff
@@ -92,6 +99,25 @@ def get_fork_point_commit(self) -> Optional["Commit"]:
         else:
             return Commit.first(sha=self.fork_point_sha, repository=self.repository)
 
+    @cached_property
+    def on_default_branch(self) -> bool:
+        """
+        Note(JP): weird: we have user-given data about whether a benchmark
+        result has been obtained for a commit on the default branch or not (In
+        the commit info object that a user can provide, we specify that "Set
+        this to `null` or leave this out to indicate that this benchmark result
+        has been obtained for the default branch."). But here we don't make
+        this easily accessible.
+
+        And then we have (potentially?) GitHub's view on whether a commit is on
+        the default branch or not. But we also don't make this easily
+        accessible.
+
+        This method is an attempt to make "this" easily accessible. But: what
+        if both are in conflict? Can they be?
```

Comment:
A reply to this docstring 🙂 

`sha` is always user-given, and `fork_point_sha` is never user-given. The latter is always generated from the GitHub HTTP API response when we `get_github_commit_metadata()`. So this new property is guaranteed to represent whether the commit is truly on the default branch or not, according to GitHub.

The user-given `branch` and `pr_number` fields are a little trickier. The `pr_number` field (the documentation of which contains the line you quoted above) is ONLY used to autofill the `branch` field if it's not given. Then the `pr_number` is discarded, not stored on the Commit object. So what do we use `branch` for, knowing that it's user-given?

It was initially supposed to be the main way to "check if the commit is on the default branch", but there were two problems: (1) getting the actual default branch name for a given repo from the GitHub HTTP API is slightly expensive/unsafe and we don't cache/store that anywhere, and (2) of course the branch is user-given so we can't fully trust that there's not an error. So I landed on "sha == fork_point_sha" as a better way to make this check.

Where else do we use `branch`? Probably the most important is in the [commit ancestry query](https://github.com/conbench/conbench/blob/03a9c846fc734e2b9bba3b143aaaabf34fd0c96e/conbench/entities/commit.py#L263), when a commit is on a PR branch and we want to find other commits on the same PR branch. Other than that it's basically only displayed in the UI.

So to answer your question: what if they are in conflict? After reviewing the code, I don't think anything bad would happen except that the UI might display confusing branch names.

Reply:
Thank you!!

Reply:
> is always generated from the GitHub HTTP API response when we get_github_commit_metadata()

Yes, except for "always" not being true, as of https://github.com/conbench/conbench/issues/904.

In that case we could fall back to trusting user input, but instead it's probably better to improve on #904.

Reply:
> So what do we use branch for, knowing that it's user-given?

Thanks for elaborating on this. This is helpful, and will be helpful again in the future.

> So I landed on "sha == fork_point_sha" as a better way to make this check.

I certainly like that today we derive this "on_default_branch" property from
- the commit hash (user-given, but 100 % _trusted_ (somewhere we need to start in terms of trust))
- repository metadata (not user-given, indirectly user-given via repo spec/URL)

> Where else do we use branch? Probably the most important is in the [commit ancestry query](https://github.com/conbench/conbench/blob/03a9c846fc734e2b9bba3b143aaaabf34fd0c96e/conbench/entities/commit.py#L263), when a commit is on a PR branch and we want to find other commits on the same PR branch.

That's interesting! Thanks for clarification.

A few points, thinking out loud:
- A commit can be on more than one branch (correct?). That is the reason for why only the commit hash as user-given input is not sufficient to do branch-local analysis. That is, the user _must_ provide pr number or branch, fundamentally. Correct?
- When the user provides an erroneous branch name or PR number: nothing we can do; that's fully OK (responsibility boundary thinking).

I think we could say "we want to trust user-given branch information". In fact, we _must_ trust user-given branch information. We can/should tell users already today that if they do not provide correct pr_number/branch information then quite a bit of analysis is broken/missed out on. We _could_ conclude from here that we are in fact already going all in on this piece of information.

Then we could do the "on default branch?" check based on the branch name, and this might be a little easier to maintain / reason about.

More brainstorming, not really important, and maybe stupid:

- Make individual benchmark result have `repo_url` and `branch` properties.
- Store repository entity in the DB. A repository entity has a default branch name. It's mutable, but changes rarely. What if it changes? We may try to do a dynamic lookup via GitHub HTTP API upon repo initialization (when it's not yet known), and then every now and then (e.g., upon container start).
- Cache in memory: repo_url->default_branch (DB lookup only needed during app init).

Reply:
These are all really good thoughts, thank you! I remember thinking about some of these things before but didn't write them down. 

>In fact, we _must_ trust user-given branch information.

I really like this statement. In practice, I remember it being difficult to get it to work from the start. When setting up various CI code that reports to Conbench, there were a handful of times that the environment variables weren't set up correctly or something else went wrong, and I ended up "polluting" the Conbench databases with commits whose branch disagreed with `sha == fork_point_sha`. But the solution to this problem is to provide users with delightful tools (and DB schemas/validation) that result in a low probability of failure.

>A commit can be on more than one branch (correct?)

Yes, and I think that's why Conbench lives with the simplified view of git where a benchmark result is either 
- not associated with a repo,
- on the default branch,
- or is a PR commit. 

As far as I know, it's undefined what happens when a commit doesn't correspond with exactly one of those. And making this simplification assumption allows us to assume there's just one associated branch.

I like your brainstorming ideas!